### PR TITLE
CT-114: Replace block type strings with references to constants

### PR DESCRIPTION
**Pull request by ebeastlake**
## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by ebeastlake** [comment]
@mikeharv I introduced a BLOCK_TYPES constant and wanted to help with consistency by using it widely, but I ran into some scenarios where other block types aren't in the BLOCK_TYPES constant I added (and are maybe CDO only?), and I wasn't sure how far I should take this... Thoughts?

File: apps/src/applab/levels.js [hunk 1]
```diff
@@ -24,8 +25,7 @@ levels.simple = {
         <value name="ID"><block type="text"><title name="TEXT">id</title></block></value> \
         <value name="HTML"><block type="text"><title name="TEXT">html</title></block></value></block>'
   ),
-  startBlocks:
-    '<block type="when_run" deletable="false" x="20" y="20"></block>',
+  startBlocks: `<block type="${BLOCK_TYPES.whenRun}" deletable="false" x="20" y="20"></block>`,
```
Comment (mikeharv) [inline apps/src/applab/levels.js:28]: I can't begin to guess why we'd be listing Blockly stuff in App Lab startBlocks...

File: apps/src/block_utils.js [hunk 2]
```diff
@@ -238,7 +239,7 @@ exports.forceInsertTopBlock = function (input, blockType) {
     if (child.nodeType === 1) {
       var type = child.getAttribute('type');
       if (
-        type !== 'procedures_defnoreturn' &&
+        type !== BLOCK_TYPES.procedureDefinition &&
```
Comment (mikeharv) [inline apps/src/block_utils.js:242]: This whole file could be used with either version of Blockly.

File: apps/src/blockly/customBlocks/cdoBlockly/spritelabBlocks.js [hunk 3]
```diff
@@ -241,19 +242,21 @@ export const blocks = {
...
     generator.behavior_definition = generator.procedures_defnoreturn;
 
-    Blockly.Procedures.DEFINITION_BLOCK_TYPES.push('behavior_definition');
+    Blockly.Procedures.DEFINITION_BLOCK_TYPES.push(
+      BLOCK_TYPES.behaviorDefinition
+    );
     Blockly.Variables.registerGetter(
       Blockly.BlockValueType.BEHAVIOR,
-      'gamelab_behavior_get'
+      BLOCK_TYPES.behaviorGet
```
Comment (mikeharv) [inline apps/src/blockly/customBlocks/cdoBlockly/spritelabBlocks.js:259]: The changes here are probably fine, but I think we'd also want to delete this file sometime after the migration is completed.

File: apps/src/dance/blocks.js [hunk 4]
```diff
@@ -113,9 +114,9 @@ export default {
         FUNCTION_NAME_LABEL: i18n.behaviorEditorLabel(),
         FUNCTION_DESCRIPTION_LABEL: i18n.behaviorEditorDescription(),
       },
-      'behavior_definition',
+      BLOCK_TYPES.behaviorDefinition,
```
Comment (mikeharv) [inline apps/src/dance/blocks.js:117]: Fun fact: Behaviors as a concept were originally intended for use in both Dance and Sprite Lab, but we never released anything using them in the former. IMHO, we should consider deleting any code that references them here.

File: apps/src/feedback.js [hunk 5]
```diff
@@ -762,7 +763,7 @@ FeedbackUtils.prototype.getFeedbackMessage = function (options) {
         var hasWhenRun = Blockly.mainBlockSpace
           .getTopBlocks()
           .some(function (block) {
-            return block.type === 'when_run' && block.isUserVisible();
+            return block.type === BLOCK_TYPES.whenRun && block.isUserVisible();
```
Comment (mikeharv) [inline apps/src/feedback.js:766]: Unsure which labs use this.

File: apps/src/feedback.js [hunk 6]
```diff
@@ -1412,7 +1413,10 @@ FeedbackUtils.prototype.checkForEmptyContainerBlockFailure_ = function () {
   }
 
   const type = emptyBlock.type;
-  if (type === 'procedures_defnoreturn' || type === 'procedures_defreturn') {
+  if (
+    type === BLOCK_TYPES.procedureDefinition ||
+    type === 'procedures_defreturn'
```
Comment (mikeharv) [inline apps/src/feedback.js:1418]: This feels inconsistent, but how many string constants to we want to make?

File: apps/src/level_base.js [hunk 7]
```diff
@@ -1,4 +1,5 @@
 // Functions for checking required blocks.
+import {BLOCK_TYPES} from '@cdo/apps/blockly/constants';
```
Comment (mikeharv) [inline apps/src/level_base.js:2]: AFAIK, the "required blocks" feature a) isn't used in any new curricula and b) might not even work right any more.

File: apps/src/music/analytics/AnalyticsReporter.ts [hunk 8]
```diff
@@ -7,6 +7,7 @@ import {
   flush,
 } from '@amplitude/analytics-browser';
 import Lab2MetricsReporter from '@cdo/apps/lab2/Lab2MetricsReporter';
+import {BLOCK_TYPES} from '@cdo/apps/blockly/constants';
```
Comment (mikeharv) [inline apps/src/music/analytics/AnalyticsReporter.ts:10]: No idea what this file is for...

File: apps/src/p5lab/P5Lab.js [hunk 9]
```diff
@@ -921,7 +922,7 @@ export default class P5Lab {
           var allBlocks = Array.from(xml.querySelectorAll('xml > block'));
           var toRemove = allBlocks.filter(element => {
             return (
-              element.getAttribute('type') === 'behavior_definition' &&
+              element.getAttribute('type') === BLOCK_TYPES.behaviorDefinition &&
```
Comment (mikeharv) [inline apps/src/p5lab/P5Lab.js]: This is related to saving Sprite Lab toolbox blocks as XML.

File: apps/src/turtle/requiredBlocks.js [hunk 10]
```diff
@@ -4,6 +4,7 @@
  * In the future, some of these may be moved to common.js.
  */
 
+import {BLOCK_TYPES} from '@cdo/apps/blockly/constants';
```
Comment (mikeharv) [inline apps/src/turtle/requiredBlocks.js:7]: Turtle is the name for "Artist' lab. See comment above about required blocks.

**Review by mikeharv — COMMENTED** [review]
This definitely feels like it could take us down a rabbit hole!

**Pull request closed without merging**